File: core/src/ru/myitschool/dcrawler/ai/pathfinding/graph/RoomNode.java
package ru.myitschool.dcrawler.ai.pathfinding.graph;

import ru.myitschool.dcrawler.dungeon.Exit;
import ru.myitschool.dcrawler.dungeon.Room;

public class RoomNode extends Node {

    private Room room;

    public RoomNode(int x, int y, Room room) {
        super(x, y);
        this.room = room;
    }

    @Override
    public boolean canConnect(Node node) {
        boolean exit = false;
        if (this.getX() - 1 == node.getX() && this.getY() == node.getY()) {
            exit = room.hasExit(Exit.DIRECTION_WEST);
        }
        if (this.getX() + 1 == node.getX() && this.getY() == node.getY()) {
            exit = room.hasExit(Exit.DIRECTION_EAST);
        }
        if (this.getX() == node.getX() && this.getY() - 1 == node.getY()) {
            exit = room.hasExit(Exit.DIRECTION_NORTH);
        }
        if (this.getX() == node.getX() && this.getY() + 1 == node.getY()) {
            exit = room.hasExit(Exit.DIRECTION_SOUTH);
        }
        return exit;
    }

    @Override
    public int connectionCost(Node to) {
        return 1;
    }

    @Override
    public boolean isReachable() {
        return true;
    }
}

File: core/src/com/wolg_vlad/dcrawler/entities/ai/pathfinding/graph/RoomNode.kt
package com.wolg_vlad.dcrawler.entities.ai.pathfinding.graph

import com.wolg_vlad.dcrawler.dungeon.Exit
import com.wolg_vlad.dcrawler.dungeon.Room

class RoomNode(x: Int, y: Int, private val room: Room) : Node(x, y) {
    override fun canConnect(node: Node): Boolean {
        var exit = false
        if (x - 1 == node.x && y == node.y) {
            exit = room.hasExit(Exit.DIRECTION_WEST)
        }
        if (x + 1 == node.x && y == node.y) {
            exit = room.hasExit(Exit.DIRECTION_EAST)
        }
        if (x == node.x && y - 1 == node.y) {
            exit = room.hasExit(Exit.DIRECTION_NORTH)
        }
        if (x == node.x && y + 1 == node.y) {
            exit = room.hasExit(Exit.DIRECTION_SOUTH)
        }
        return exit
    }

    override fun connectionCost(to: Node?): Int {
        return 1
    }

    override val isReachable: Boolean
        get() = true

}